File: LevelAndroidSDK/sdk/src/main/java/com/theshopatvsp/levelandroidsdk/ble/model/response/Frame.java
package com.theshopatvsp.levelandroidsdk.ble.model.response;

import  com.theshopatvsp.levelandroidsdk.ble.helper.BitsHelper;
import  com.theshopatvsp.levelandroidsdk.ble.model.constants.LevelColor;
import  com.theshopatvsp.levelandroidsdk.ble.model.constants.LevelModel;
import  com.theshopatvsp.levelandroidsdk.ble.model.response.DataPacket;

/**
 * Created by andrco on 5/12/16.
 */
public class Frame extends DataPacket {
    private LevelModel model;
    private LevelColor color;

    public Frame(LevelModel model, LevelColor color) {
        this.model = model;
        this.color = color;
    }

    public Frame(byte[] packet) {
        if (packet != null && packet.length > 2) {
            int m = BitsHelper.convertTo16BitInteger(packet[3], packet[2]);
            int c = BitsHelper.convertTo16BitInteger(packet[5], packet[4]);

            this.model = LevelModel.getByNumber(m);
            this.color = LevelColor.getByNumber(c);
        }
    }

    public LevelModel getModel() {
        return model;
    }

    public LevelColor getColor() {
        return color;
    }

    public byte[] getPacket() {
        byte bytes[] = new byte[4];

        byte modelBytes[] = BitsHelper.convertIntTo2Bytes(model.getModelNumber());
        byte colorBytes[] = BitsHelper.convertIntTo2Bytes(color.getColorNumber());

        bytes[0] = modelBytes[0];
        bytes[1] = modelBytes[1];
        bytes[2] = colorBytes[0];
        bytes[3] = colorBytes[1];

        return bytes;
    }
}

File: LevelIOSSdkExample/LevelIOSSdkExample/LevelIOSSdk/model/Frame.swift
//
//  Frame.swift
//  LevelIOSSDK
//
//  Created by Andrew Cook on 6/24/16.
//  Copyright © 2016 TheShop. All rights reserved.
//

import Foundation

class Frame: DataPacket {
    var model: LevelModel
    var color: LevelColor
    
    override init() {
        self.model = .NotSet
        self.color = .NotSet
        
        super.init();
    }
    
    convenience init(model: LevelModel, color: LevelColor) {
        self.init()
        
        self.model = model
        self.color = color
    }
    
    convenience init(bytes: [UInt8]) {
        self.init()
        
        let modelNum = Int(BitsHelper.convertToUInt16(msb: bytes[3], lsb: bytes[2]))
        let colorNum = Int(BitsHelper.convertToUInt16(msb: bytes[5], lsb: bytes[4]))
        
        self.model = LevelModel(rawValue: modelNum)!
        self.color = LevelColor(rawValue: colorNum)!
    }
    
    override func getPacket() -> [UInt8] {
        var bytes: [UInt8] = [UInt8]()
        
        var modelBytes: [UInt8] = BitsHelper.convertTo2Bytes(number: model.rawValue)
        var colorBytes: [UInt8] = BitsHelper.convertTo2Bytes(number: color.rawValue)
        
        bytes.append(modelBytes[0])
        bytes.append(modelBytes[1])
        bytes.append(colorBytes[0])
        bytes.append(colorBytes[1])
        
        return bytes
    }
}
